### PR TITLE
Handle unexpected behavior with URI#merge and subpaths missing trailing slashes

### PR DESCRIPTION
**Pull request by drcapulet**
Due to how `URI#merge` works (which `URI#+` is aliased to), when using a source under a subpath, it can get overwritten:

```ruby
> URI.parse('https://gems.example.com/private') + './api/v1/dependencies'
=> #<URI::HTTPS https://gems.example.com/api/v1/dependencies>
> URI.parse('https://gems.example.com/private/') + './api/v1/dependencies'
=> #<URI::HTTPS https://gems.example.com/private/api/v1/dependencies>
```

It's not immediately obvious that errors fetching would be caused by a missing slash, especially if you have a server mounted at both `/` and `/private`.

This PR makes sure that calls to `URI#+` are more obvious and subpaths aren't overwritten and instead appended.

Fixes https://github.com/rubygems/rubygems/issues/1829.

### CODE REVIEW

**Comment by bronzdoc** [comment]
Hey @drcapulet! could you add the test into `test/rubygems/test_gem_source.rb`? thank you!

**Comment by drcapulet** [comment]
@bronzdoc Was there something you'd want me to cover in `test/rubygems/test_gem_source.rb` that isn't in `test/rubygems/test_gem_source_subpath_problem.rb`? Or do you mean try merging them together? If so, I wasn't sure on a good way to achieve what I wanted with `#setup`.

**Comment by bronzdoc** [comment]
@drcapulet sorry for the confusion. I meant "move" not "add"

**Comment by hsbt** [comment]
@drcapulet Hi, Can you move the your test into the `test_gem_source.rb`?

After that, I will merge this.

**Comment by drcapulet** [comment]
@hsbt As I mentioned earlier, I wasn't sure of a good way to set `@gem_repo` - do you just want me to copy the setup across the three tests?

**Comment by deivid-rodriguez** [comment]
@hsbt Should @drcapulet copy the test setup across each of the tests or do you have another preferred way? In my opinion, a separate test file like it is coded currently is fine.

By the way, does this fix or is related to https://github.com/rubygems/rubygems/issues/1829?

**Comment by drcapulet** [comment]
@deivid-rodriguez Yes, this should fix #1829.

**Comment by deivid-rodriguez** [comment]
Great! Could you rebase this PR to make sure everything is still good, and so that the PR is ready to merge in case @hsbt is fine with it now?

File: lib/rubygems/source.rb
Comment (deivid-rodriguez) [inline lib/rubygems/source.rb:83]: To make the actual fix here more apparent, could we extract this logic to a `enforce_trailing_slash` helper, and keep the previous code untouched otherwise?

```suggestion
    bundler_api_uri = enforce_trailing_slash(uri) + './api/v1/dependencies'
```

**Review by deivid-rodriguez — APPROVED** [review]
I rebased this PR and added my suggestions. I'll merge in a couple of days unless there's further feedback.

**Comment by deivid-rodriguez** [comment]
Thanks @drcapulet!

**Pull request merged**
